### PR TITLE
Add hidden field

### PR DESCRIPTION
**Pull request by CristhianMotoche**
Solves #2 

# Changes
- Add `Hidded`, which is a new constructor for `KeyValue`
- Ignoring any Hidden attribute by returning `(VNull, undefined)`
- `undefined` won't be evaluated because its filtered from the KeyValue list

Please, let me know if this would solve the issue. :smiley:

### CODE REVIEW

**Comment by moleike** [comment]
Hi Cristhian, thanks for your patch! I am reviewing this later 👍

**Review by moleike — CHANGES_REQUESTED** [review]
Very impressive work!
Please see my comments, and also:
You could change the evalObj method to account for hidden fields, then in the VObj constructor change the keys from `Text` to a new type e.g. `data Key = Key Text | Hidden Text`, the idea here is that fields should all be usable, the only time we care about whether it's a hidden field is when we are materializing (Value -> JSON) or checking equality.

If this is not clear, please let me know!

File: src/Language/Jsonnet/Common.hs [hunk 1]
```diff
@@ -65,7 +65,7 @@ instance Alpha UnyOp
 data KeyValue a = KeyValue
   { key :: a,
     value :: a
-  }
+  } | Hidden { key :: a }
```
Comment (moleike) [inline src/Language/Jsonnet/Common.hs:68]: Hidden fields, also have values, e.g.
```hs
{ foo : "bar", baz(x) :: x + 1 }
```
In this case, you might use the baz method in the program, regardless of the object being serialized to JSON.

Another example, the functions in the standard library are all hidden fields of the std object (https://jsonnet.org/ref/stdlib.html)

Comment (moleike) [reply src/Language/Jsonnet/Common.hs:68]: My suggestion here would be to add an additional field `hidden :: Bool`.

**Comment by CristhianMotoche** [comment]
Thanks for the suggestions @moleike I'll review them later today.

**Comment by CristhianMotoche** [comment]
Hi @moleike I tried your approach but I got stuck with this type error:
``hs
...l/jsonnet-haskell/src/Language/Jsonnet/Std.hs:42:26: error:
    • No instance for (Integral Key) arising from a use of ‘inj’
    • In the expression: inj (H.keys @Key @Value)
      In the expression: ("objectFields", inj (H.keys @Key @Value))
      In the expression:
        [("assertEqual", inj assertEqual), ("type", inj valueType),
         ("isString", inj (isType "string")),
         ("isBoolean", inj (isType "boolean")), ....]
   |
42 |         ("objectFields", inj (H.keys @Key @Value)),

```

Please, let me know if I'm going in the right direction or I lost at some point. I understood that `manifest` makes the final conversion of `Value` to `JSON` and it's used in `toString` and `equalTo`.

**Comment by CristhianMotoche** [comment]
I managed to fix the error type, but I'm not sure about my implementation.

**Comment by moleike** [comment]
Hi @CristhianMotoche, the problem you encounter is due to the impl. of `objectFields` method. So you don't need an `HasValue` instance for `Key`:

```diff
 diff --git a/src/Language/Jsonnet/Eval.hs b/src/Language/Jsonnet/Eval.hs
index 8dc0428..d0c0993 100644
--- a/src/Language/Jsonnet/Eval.hs
+++ b/src/Language/Jsonnet/Eval.hs
@@ -124,10 +124,6 @@ instance HasValue Double where
   proj v = throwTypeMismatch "number" v
   inj = VNum
 
-instance HasValue Key where
-  proj = undefined
-  inj = undefined
-
 instance {-# OVERLAPS #-} Integral a => HasValue a where
   proj (VNum n) = pure (round n)
   proj v = throwTypeMismatch "number" v
diff --git a/src/Language/Jsonnet/Std.hs b/src/Language/Jsonnet/Std.hs
index ba2a1d3..2f8d67a 100644
--- a/src/Language/Jsonnet/Std.hs
+++ b/src/Language/Jsonnet/Std.hs
@@ -9,6 +9,7 @@ import Control.Applicative
 import Control.Monad.Except
 import qualified Data.ByteString as B
 import Data.Foldable
+import Data.HashMap.Lazy (HashMap)
 import qualified Data.HashMap.Lazy as H
 import Data.List
 import Data.Text (Text)
@@ -39,7 +40,7 @@ std = VObj $ (Thunk . pure) <$> H.fromList (map (\(k, v) -> (Visible k, v)) xs)
         ("isObject", inj (isType "object")),
         ("isArray", inj (isType "array")),
         ("isFunction", inj (isType "function")),
-        ("objectFields", inj (H.keys @Key @Value)),
+        ("objectFields", inj objectFields),
         ("length", inj length'),
         ("abs", inj (abs @Double)),
         ("sign", inj (signum @Double)), -- incl. 0.0, (-0.0), and NaN
@@ -98,6 +99,10 @@ std = VObj $ (Thunk . pure) <$> H.fromList (map (\(k, v) -> (Visible k, v)) xs)
         ("reverse", inj (reverse @Value))
       ]
 
+
+objectFields :: HashMap Key Value -> [Text]
+objectFields o = [k | Visible k <- H.keys o]
+
 assertEqual :: Value -> Value -> Eval Bool
 assertEqual a b = do
   a' <- manifest a
```

**Comment by moleike** [comment]
> I understood that manifest makes the final conversion of Value to JSON and it's used in toString and equalTo. Therefore, I should filter the visible keys there.

That's correct.
> Please, let me know if I'm going in the right direction

Yes you are 👍

**Comment by CristhianMotoche** [comment]
Thank you so much for you kind guidance @moleike Let me know if you have any other comment/suggestion. :smiley:

**Comment by moleike** [comment]
@CristhianMotoche I think the issue is solved with your changes. Congrats on tackling this issue!

**Review by moleike — APPROVED** [review]
LGTM

**Comment by moleike** [comment]
There's one minor issue remaining, but it can be fixed later on:
The parses will fail to parse `{ foo(x) :: x }`, which is sugar for `{ foo :: function(x) x }`.

**Pull request merged**